TESTS/mbed_drivers/timeout/timeout_tests.h

```diff
@@ -0,0 +1,416 @@
+}
+
+#if DEVICE_SLEEP
+/** Template for tests: timeout during sleep
+ *
+ * Test timeout during sleep
+ * Given a Timeout object with a callback attached with @a attach()
+ *     and the uC in a sleep mode
+ * When given delay elapses
+ * Then the callback is called
+ *     and elapsed time matches given delay
+ *
+ * Test timeout during sleep
+ * Given a Timeout object with a callback attached with @a attach_us()
+ *     and the uC in a sleep mode
+ * When given delay elapses
+ * Then the callback is called
+ *     and elapsed time matches given delay
+ */
+template<typename T, us_timestamp_t delay_us, us_timestamp_t delta_us>
+void test_sleep(void)
+{
+    Semaphore sem(0, 1);
+    T timeout;
+    Timer timer;
+
+    timer.start();
+    sleep_manager_lock_deep_sleep();
+    timeout.attach_callback(mbed::callback(sem_callback, &sem), delay_us);
+
+    bool deep_sleep_allowed = sleep_manager_can_deep_sleep();
+    TEST_ASSERT_FALSE_MESSAGE(deep_sleep_allowed, "Deep sleep should be disallowed");
+    sleep();
+
+    int32_t sem_slots = sem.wait(osWaitForever);
+    timer.stop();
+    sleep_manager_unlock_deep_sleep();
+    TEST_ASSERT_EQUAL(1, sem_slots);
+    TEST_ASSERT_UINT64_WITHIN(delta_us, delay_us, timer.read_high_resolution_us());
+
+    timeout.detach();
+}
+
+#if DEVICE_LOWPOWERTIMER
+/** Template for tests: timeout during deepsleep
+ *
+ * Test timeout during deepsleep
+ * Given a LowPowerTimeout object with a callback attached with @a attach()
+ *     and the uC in a deepsleep mode
+ * When given delay elapses
+ * Then the callback is called
+ *     and elapsed time matches given delay
+ *
+ * Test timeout during deepsleep
+ * Given a LowPowerTimeout object with a callback attached with @a attach_us()
+ *     and the uC in a deepsleep mode
+ * When given delay elapses
+ * Then the callback is called
+ *     and elapsed time matches given delay
+ */
+template<typename T, us_timestamp_t delay_us, us_timestamp_t delta_us>
+void test_deepsleep(void)
+{
+    Semaphore sem(0, 1);
+    T timeout;
+    LowPowerTimer timer;
+
+    /*
+     * Since deepsleep() may shut down the UART peripheral, we wait for 10ms
+     * to allow for hardware serial buffers to completely flush.
+
+     * This should be replaced with a better function that checks if the
+     * hardware buffers are empty. However, such an API does not exist now,
+     * so we'll use the wait_ms() function for now.
+     */
+    wait_ms(10);
+
+    /*
+     * We use here the low power timer instead of microsecond timer for start and
+     * end because the microseconds timer might be disabled during deepsleep.
+     */
+    timer.start();
+    timeout.attach_callback(mbed::callback(sem_callback, &sem), delay_us);
+    /* Make sure deepsleep is allowed, to go to deepsleep */
+    bool deep_sleep_allowed = sleep_manager_can_deep_sleep();
+    TEST_ASSERT_TRUE_MESSAGE(deep_sleep_allowed, "Deep sleep should be allowed");
+    sleep();
+
+    int32_t sem_slots = sem.wait(osWaitForever);
```

Comment:
If this line is needed, then it should be moved outside timer start-stop section and wait value should be set to 0.